src/OracleService/OracleService.cs

```diff
@@ -0,0 +1,489 @@
+            }
+            else if (tx.NetworkFee + (size + tx.Attributes.GetVarSize()) * feePerByte > request.GasForResponse)
+            {
+                response.Code = OracleResponseCode.InsufficientFunds;
+                response.Result = Array.Empty<byte>();
+            }
+            size += tx.Attributes.GetVarSize();
+            tx.NetworkFee += size * feePerByte;
+
+            // Calcualte system fee
+
+            tx.SystemFee = request.GasForResponse - tx.NetworkFee;
+
+            return tx;
+        }
+
+        private void AddResponseTxSign(StoreView snapshot, ulong requestId, ECPoint oraclePub, byte[] sign, Transaction responseTx = null, Transaction backupTx = null, byte[] backupSign = null)
+        {
+            var task = pendingQueue.GetOrAdd(requestId, _ => new OracleTask
+            {
+                Id = requestId,
+                Request = NativeContract.Oracle.GetRequest(snapshot, requestId),
+                Signs = new ConcurrentDictionary<ECPoint, byte[]>(),
+                BackupSigns = new ConcurrentDictionary<ECPoint, byte[]>()
+            });
+
+            if (responseTx != null)
+            {
+                task.Tx = responseTx;
+                var data = task.Tx.GetHashData();
+                task.Signs.Where(p => !Crypto.VerifySignature(data, p.Value, p.Key)).ForEach(p => task.Signs.Remove(p.Key, out _));
+            }
+            if (backupTx != null)
+            {
+                task.BackupTx = backupTx;
+                var data = task.BackupTx.GetHashData();
+                task.BackupSigns.Where(p => !Crypto.VerifySignature(data, p.Value, p.Key)).ForEach(p => task.BackupSigns.Remove(p.Key, out _));
+                task.BackupSigns.TryAdd(oraclePub, backupSign);
+            }
+            if (task.Tx == null)
+            {
+                task.Signs.TryAdd(oraclePub, sign);
+                task.BackupSigns.TryAdd(oraclePub, sign);
+                return;
+            }
+
+            if (Crypto.VerifySignature(task.Tx.GetHashData(), sign, oraclePub))
+                task.Signs.TryAdd(oraclePub, sign);
+            else if (Crypto.VerifySignature(task.BackupTx.GetHashData(), sign, oraclePub))
+                task.BackupSigns.TryAdd(oraclePub, sign);
+            else
+                throw new RpcException(-100, "Invalid response transaction sign");
+
+            if (CheckTxSign(snapshot, task.Tx, task.Signs) || CheckTxSign(snapshot, task.BackupTx, task.BackupSigns))
+            {
+                finishedCache.TryAdd(requestId, new DateTime());
+                pendingQueue.TryRemove(requestId, out _);
+            }
+        }
+
+        private static byte[] Filter(string input, string filterArgs)
+        {
+            if (string.IsNullOrEmpty(filterArgs))
+                return Utility.StrictUTF8.GetBytes(input);
+
+            NJObject beforeObject = NJObject.Parse(input);
+            NJArray afterObjects = new NJArray(beforeObject.SelectTokens(filterArgs));
```

Comment:
Is `SelectTokens()` deterministic?

Reply:
No explicit description. I think it should be deterministic, as it supports array indexes. 
https://www.newtonsoft.com/json/help/html/SelectToken.htm